src/fundamental/macro-fundamental.h

```diff
@@ -50,6 +50,15 @@
 #define UNIQ_T(x, uniq) CONCATENATE(__unique_prefix_, CONCATENATE(x, uniq))
 #define UNIQ __COUNTER__
 
+#define ONCE __ONCE(UNIQ_T(_once_, UNIQ))
+#define __ONCE(o)                            \
+        ({                                   \
+                static bool (o) = true;      \
+                bool __o = (o);              \
+                (o) = false;                 \
+                __o;                         \
```

Comment:
hmm, i am a bit concerned about thread-safety for this. generic functionality like this should have some superficial concern about that. i mean, we use thread very conservatively, but we do use them, and we log from threads, and in logging you want to use this.

I figure you could easily rewrite this to use gcc's `__sync_bool_compare_and_swap` builtin, which is universally supported on llvm too these days, and does pretty much exactly what you want to do here..

note that even then this is a noticably weaker model than pthread_once() even though similarly named: pthread_once will wait until the code that wants to run once is finished before allowing any other users to continue. Now I don't think those semantics make sense for our usecase here, hence I would not be too concerned about this asymmetry. however, i think a brief comment should be added here that explains the difference. Maybe something like this:

```c
/* Note that this works differently from pthread_once(): this macro does not synchronize code execution, i.e. code that is run conditionalized on this macro will run concurrently to all other code conditionalized the same way, there's no ordering or completion enforced. */
```

or something like that

Reply:
what about using #include <stdatomic.h> functions? I believe gcc and llvm both support the C11 standard atomics, so we could use atomic_flag_test_and_set()

Reply:
that would mean we wouldn't work with anything older than gcc 4.9 (~2016), and clang 3.6 (~2015). 

The question is if we care, i.e. if any of our CIs or relevant distros or so use an older compiler than that? @mrc0mmand do you know?

Reply:
(quite frankly, i#d just use the __sync_bool_compare_and_swap() stuff. it's a lot better supported without any doubt. and we already use it at various places)

Reply:
> that would mean we wouldn't work with anything older than gcc 4.9 (~2016), and clang 3.6 (~2015).
> 
> The question is if we care, i.e. if any of our CIs or relevant distros or so use an older compiler than that? @mrc0mmand do you know?

IIRC the last problematic job was CentOS 7 (with gcc 4.x), but now even Bionic runs gcc 7/clang 6 and CentOS 8 runs gcc 8/clang 10, so that, __theoretically__ shouldn't be an issue.

Reply:
ack, i'll switch to using __sync_bool_compare_and_swap()